src/systests/test_userbehavior.py

```diff
@@ -0,0 +1,116 @@
+# (at your option) any later version.
+#
+# SOCBED is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with SOCBED. If not, see <http://www.gnu.org/licenses/>.
+
+import json
+import os
+import re
+import time
+from http.client import HTTPConnection, CannotSendRequest
+from collections import namedtuple
+
+import pytest
+import requests
+
+from attacks.ssh import SSHTargets
+from systests.helpers import try_until_counter_reached
+from vmcontrol.sessionhandler import SessionHandler
+from vmcontrol.vmmcontroller import VBoxController
+
+MAX_RUNTIME = 10 * 60  # Ten minutes
+pytestmark = [pytest.mark.systest]
+
+
+@pytest.fixture(scope="module")
+def session():
+    sh = SessionHandler(VBoxController())
+    sh.start_session()
+    yield
+    sh.close_session()
+
+
+@pytest.fixture(scope="module")
+def timeout_counter():
+    return time.perf_counter() + MAX_RUNTIME
+
+
+@pytest.mark.usefixtures("session", "timeout_counter")
+class TestUserbehavior:
+    log_server = SSHTargets.log_server
+
+    def test_browsing(self, timeout_counter):
+        elasticsearch_query_values = namedtuple(
+            "elasticsearch_query_values", "value search_keyword timestamp_key")
+        called_website = "www.general-anzeiger-bonn.de"
+        es_query = elasticsearch_query_values(called_website, "dns.question.name", "@timestamp")
+        self.check_remote_log_format(es_query, timeout_counter)
+
+    def check_remote_log_format(self, es_query, timeout_counter):
+        try_until_counter_reached(
+            lambda: self.query_elasticsearch(es_query, timeout_counter),
+            timeout_counter,
+            exception=IndexError,
+            assertion_func=lambda x: self.check_elasticsearch_response(x, es_query))
+
+    def is_iso8601_timestamp(self, timestamp):
+        iso8601_a = re.compile(
+            r"\d\d\d\d-\d\d-\d\dT\d\d:\d\d:\d\d.\d\d\d\d\d\d[+-]\d\d:\d\d")
+        iso8601_b = re.compile(
+            r"\d\d\d\d-\d\d-\d\dT\d\d:\d\d:\d\d.\d\d\dZ")
+        return (iso8601_a.match(timestamp) is not None
+                or iso8601_b.match(timestamp) is not None)
+
+    def test_is_iso8601_timestamp(self):
+        valid_timestamps = [
+            "2017-06-01T15:45:15.495964+02:00"
+            "2021-02-13T05:50:01.109Z"]
+        invalid_timestamps = [
+            "May 31 16:03:00",
+            "May",
+            ""]
+        assert all(
+            self.is_iso8601_timestamp(valid_timestamp)
+            for valid_timestamp in valid_timestamps)
+        assert not any(
+            self.is_iso8601_timestamp(invalid_timestamp)
+            for invalid_timestamp in invalid_timestamps)
+
+    def query_elasticsearch(self, es_query, timeout_counter, conn=-1):
```

Comment:
I find it a bit confusing and dangerous to check the cluster health in order to make sure Elasticsearch is up. The subsequent HTTPConnection could still fail for other reasons. Isn't it possible to add lines 98-99 to the subsequent try_until_counter_reached (as a function) and omit the health check?

Reply:
I'm not sure I fully understand; the cluster being ready fully implies that Elasticsearch is available, which in turn implies that it is also reachable. Given that the health check uses requests and the following part uses http.client, it should work every time because requests itself uses http.client at a lower level.

I mean, I would just slap the `conn = HTTPConnection(f"{self.log_server.hostname}:9200")` part into the `try_until_counter_reached` function at line 101, but I honestly dont know how. I need the the return value `conn` for later, which would be theoretically possible with a standalone lambda expression, but since it is enclosed in our own helper function, I'm not really sure how to properly accomplish this. Making it global seems like a weird idea.

Reply:
I just saw that `query_elasticsearch` is already called in a `try_until_counter_reached` function (line 61) and should therefore not use `try_until_counter_reached` by itself. The function should simply look like this (not tested!):
```python
    def query_elasticsearch(self, es_query):
        headers = {"Content-type": "application/json", "Accept": "text/plain"}
        search_query = json.dumps({"query": {"constant_score": {"filter": {"term": {
            es_query.search_keyword: es_query.value}}}}})
        conn = HTTPConnection(f"{self.log_server.hostname}:9200")
        conn.request(method="POST", url="/_search", body=str(search_query), headers=headers)
        return str(conn.getresponse().read(), 'utf-8')
```
You probably need to add the exceptions from line 106 to line 63 for this to work.

Reply:
Implemented and tested, seems to work perfectly